### DISCUSSION TITLE
for_display_only? What's going on?

### DISCUSSION THREAD

**Discussion by BilldarBagdar** [markusschloesser/MackieC4_P3#14] [Q&A] [answered]
A:  That variable name (currently) found 20 times in EncoderController is used in two ways but is ultimately IMHO (currently) a bit of a distracting illusion (a red herring?)

It appears once as a method parameter name in the definition: 
`def __reassign_encoder_parameters(self, for_display_only):`
and IIRC, it appears everywhere else as a local variable name where `__reassign_encoder_parameters(for_display_only=False)` is called from "other code".

When a method is defined as depicted above for_display_only is treated as a required parameter whose value must be provided when the method is called.  However, you can define methods with parameters that will use default values:
`def __reassign_encoder_parameters(self, for_display_only=False):`
for_display_only here is treated as an optional parameter whose value only needs to be provided when the method is called and the non-default behavior is desired (when for_display_only = True behavior is desired)

How it is actually coded in this case depends on the programmer, because the logic works either way.  We as the programming team and ultimately you as the repo owner get to decide.  Is it going to be rare that we would want non-default display behavior from the method?  If we define the method with `for_display_only=False` as the default now, then we can remove `for_display_only=False` from all of the 19 method calls.  But if, for example, around half of those 19 calls would be calls using the non-default value anyway, the utility of defining the method "with default" drops dramatically.

...and now come the fish.  for_display_only is not (currently) referenced inside `__reassign_encoder_parameters(self, for_display_only)`.  I guess Leigh had plans for it but never got to implementing those plans.  You can tell because PyCharm should decorate the for_display_only text in a grey Font (the method definition "signature" parameter name only, line 709) because it is declared but not referenced.  (grey text like that is standard PyCharm behavior.  when you first declare a new variable name it will appear as grey text until you "use it" somewhere. (new variable in this sense just means new inside the current "indentation scope".))

**Comment by BilldarBagdar** [comment]
> it will appear as grey text until you "use it" somewhere

Actually, the "grey text" I see is probably dependent on the fact that I don't use PyCharm's "dark mode" with the black backgrounds, and I haven't messed with any defaults.  Your mileage could easily vary...

**Reply by markusschloesser** [reply]
1. `for_display_only=False` also appears 12 times in the MackieControl scripts, both for Live 10 and 11.
2. in the MackieControl scripts `for_display_only` is actually used in  
    `

    def __reassign_channel_strip_parameters(self, for_display_only):
        display_parameters = []
        for s in self._ChannelStripController__channel_strips:
            vpot_param = (None, None)
            slider_param = (None, None)
            vpot_display_mode = VPOT_DISPLAY_SINGLE_DOT
            slider_display_mode = VPOT_DISPLAY_SINGLE_DOT
            if self._ChannelStripController__assignment_mode == CSM_VOLPAN:
                if s.assigned_track():
                    if s.assigned_track().has_audio_output:
                        vpot_param = (
                         s.assigned_track().mixer_device.panning, 'Pan')
                        vpot_display_mode = VPOT_DISPLAY_BOOST_CUT
                        slider_param = (
                         s.assigned_track().mixer_device.volume, 'Volume')
                        slider_display_mode = VPOT_DISPLAY_WRAP
                elif self._ChannelStripController__assignment_mode == CSM_PLUGINS:
                    vpot_param = self._ChannelStripController__plugin_parameter(s.strip_index(), s.stack_offset())
                    vpot_display_mode = VPOT_DISPLAY_WRAP
                    if s.assigned_track():
                        if s.assigned_track().has_audio_output:
                            slider_param = (
                             s.assigned_track().mixer_device.volume, 'Volume')
                            slider_display_mode = VPOT_DISPLAY_WRAP
                elif self._ChannelStripController__assignment_mode == CSM_SENDS:
                    vpot_param = self._ChannelStripController__send_parameter(s.strip_index(), s.stack_offset())
                    vpot_display_mode = VPOT_DISPLAY_WRAP
                    if s.assigned_track():
                        if s.assigned_track().has_audio_output:
                            slider_param = (
                             s.assigned_track().mixer_device.volume, 'Volume')
                            slider_display_mode = VPOT_DISPLAY_WRAP
                else:
                    if self._ChannelStripController__assignment_mode == CSM_IO:
                        if s.assigned_track():
                            if s.assigned_track().has_audio_output:
                                slider_param = (
                                 s.assigned_track().mixer_device.volume, 'Volume')
                    if self._ChannelStripController__flip and self._ChannelStripController__can_flip():
                        if self._ChannelStripController__any_slider_is_touched():
                            display_parameters.append(vpot_param)
                        else:
                            display_parameters.append(slider_param)
                        if not for_display_only:
                            s.set_v_pot_parameter(slider_param[0], slider_display_mode)
                            s.set_fader_parameter(vpot_param[0])
            else:
                if self._ChannelStripController__any_slider_is_touched():
                    display_parameters.append(slider_param)
                else:
                    display_parameters.append(vpot_param)
                for_display_only or s.set_v_pot_parameter(vpot_param[0], vpot_display_mode)
                s.set_fader_parameter(slider_param[0])
        self._ChannelStripController__main_display_controller.set_channel_offset(self._ChannelStripController__strip_offset())
        if len(display_parameters):
            self._ChannelStripController__main_display_controller.set_parameters(display_parameters)
`
3. How do paste code blocks properly here? Sometimes it works, most of the times it doesn't. Tried the code wrapping and spaces (actually the above code has both
4. I still haven't understood what the  `for_display_only` actually does 🙄

**Reply by BilldarBagdar** [reply]
I don't know about pasting large code blocks in this discussion thread, the backticks probably only work for one liner code quotes.  You could also just link to a bunch of code like that if it's in the repo.  [Like this](https://github.com/markusschloesser/MackieC4_P3/blob/master/EncoderController.py#L709)?

or just quote the actual usage example? (triple backticks seems to be the trick)

```
if not for_display_only:
    s.set_v_pot_parameter(slider_param[0], slider_display_mode)
    s.set_fader_parameter(vpot_param[0])
else:
    if self._ChannelStripController__any_slider_is_touched():
        display_parameters.append(slider_param)
    else:
        display_parameters.append(vpot_param)
for_display_only or s.set_v_pot_parameter(vpot_param[0], vpot_display_mode)
```
You are correct.  This example code does make use of (what appears to be a boolean valued) variable named "for_display_only".  

I suspect you already understand the fact that it's not used in the C4 script and you are just curious in general what the heck the purpose of such a curiously named variable could be.  All I can do is offer an educated guess, which is it might be about controlling "feedback".  When something is "for display only", maybe for example, only the "display" of the motorized faders should update.  The new updated values of the faders should not be sent back to Live, or some such behavior.

**Reply by markusschloesser** [reply]
> I don't know about pasting large code blocks in this discussion thread, the backticks probably only work for one liner code quotes.

It works by pasting code, then selecting and THEN clicking the code button.

> I suspect you already understand the fact that it's not used in the C4 script and you are just curious in general what the heck the purpose of such a curiously named variable could be. All I can do is offer an educated guess, which is it might be about controlling "feedback". When something is "for display only", maybe for example, only the "display" of the motorized faders should update. The new updated values of the faders should not be sent back to Live, or some such behavior.

You suspected correctly :-)
And I tend to agree, it looks like a decoupling of **moving** and **displaying** stuff, which, for encoders, makes a lot of sense. 
But then the non-usage in the C4 script makes less sense. Or is the non-usage just the "not setting" of the parameter? I think so, but please confirm .

**Reply by BilldarBagdar** [reply]
:) pasting code blocks also works here 

- by typing "triple backticks", pasting code, and typing another "triple backticks" 
- by first clicking the "Insert Code" widget (which will insert a pair of "triple backticks") and then pasting your code between the inserted "triple backtick pair".  

You have the freedom of choice. (and an implied obligation to learn which choice you prefer)

**Reply by BilldarBagdar** [reply]
I think maybe the "not setting" part might be an open question.   When a user twists a knob on the C4, it should cause the Live GUI to update and all other connected remote devices should also update their displays.  The opposite is also true.  The C4 display should update itself when the user changes something on another device and/or the Live GUI.

I suspect maybe the reason that same "midi feedback" prevention mechanism seen in the "corporate Mackie Control" remote scripts isn't used in the C4 script is because the way Leigh coded things, such feedback isn't even possible.  (But that suspicion is already the second turtle down that rabbit hole).  If processing the "receipt of" updated display information from Live (and showing updated info on display) is not connected to processing the "sending of" updated encoder information to Live, then there's no feedback chance.

What follows is just an example.  I don't know the script or the C4 device works either of these ways.  Examples only.

When a user turns an encoder on the C4 device, that "midi data" data is sent "physically over the midi cable" to Live before the script ever sees anything, and that data always reflects some physical user interaction with the device.  When Live (and/or the remote script) sends "midi data physically over the midi cable" to the C4, that data only ever updates the LEDs and LCDs.  There is an inherent disconnect between "knob turns" and "display updates".  That's one example.  The other would be when a user turns an encoder on the C4 device, the device updates its own display and then that "midi data" data is sent "physically over the midi cable" to Live.  In that case, it's possible one would need an indicator like "for display only" that would prevent/suppress sending the updated display info back "over the wire to Live" as if the user had turned that controller on the C4 (feedback).

Lot's of maybe's in that.  What I do know is that even though `for_display_only` appears 20 times in the C4 script code, it is only "used" once in the `def __reassign_encoder_parameters(self, for_display_only)` method definition; and the same thing is true in the "corporate Mackie" code you posted, that method is named `__reassign_channel_strip_parameters(self, for_display_only)` so it is easy for me to imagine myself in Leigh's shoes trying to invent a C4 script based on whatever was available for reference at the time.  I guess he started by copy/pasting the corporate Mackie code and then "hacking and slashing" his way to a working C4 script. (and it just so happens that he didn't need the `for_display_only` behavior and he didn't "clean up" loose ends after getting to a "something that works" point of satisfaction where he lost motivation for further changes)

**Reply by markusschloesser** [reply]
wait a sec, this 

> When a user twists a knob on the C4, it should cause the Live GUI to update and all other connected remote devices should also update their displays. The opposite is also true. The C4 display should update itself when the user changes something on another device and/or the Live GUI.

works fine here!
the encoder ring also moves with the knob and with the shown parameter in Live. and vice versa

**Reply by BilldarBagdar** [reply]
Yes, it does.  By describing how things "should work" I did not mean to also imply anything was not "working as it should".  More like I was just "setting the stage" for discussion that follows (not that I was thinking that hard about "staging"... )

**Reply by markusschloesser** [reply]
reading that again: Did you mean "feedback" as in an good or bad thing?
1. good: stuff is fed back to the C4 (and displayed there) (that's what i though you meant)
2. bad as in feedback loop

**Comment by BilldarBagdar** [comment]
Visual feedback to the device = good. If that data then loops back to Live = bad


This is a signature line.
edit: removed email garbage

**Comment by markusschloesser** [comment, marked as answer]
currently reading old threads and found this and now I'm pretty sure that I know the answer:

`for_display_only` in the MCU scripts is for when you only TOUCH a fader, but NOT move it. When you touch the (capacitive) fader, the display switches to showing the current fader value. When you un-touch, the display switches back to what it was showing before. 

That's the `if self._ChannelStripController__any_slider_is_touched():` part

As we don't have capacitive vpot encoders, this is not applicable to the C4

**Reply by BilldarBagdar** [reply]
Totally agree.  All references to `for_display_only` in the updated script could safely be eliminated.  The variable was used in ancient times but no longer has a valid purpose today.  (Although my MCU does not have any capacitive hardware)

"Touch" does have a meaning in the Framework code related to "Gestures" though.  "Gesture control" is what the "TouchEncoderElement" class adds to a plain "EncoderElement" in both Frameworks.  The way I understand that idea is when a user begins turning an encoder knob, for example, a "gesture" begins and the gesture doesn't end until the user stops turning the knob (or the gesture is interrupted).  Then a script can, for example, show different feedback if when a "gesture" is currently in progress.

For example, in Channel Strip mode, the "device row" encoder knobs don't do anything right? (just the encoder buttons take to you Device mode) and the "device row" display takes up both LCD rows for the mapped "device name".  If we had a reason to map the "device row" encoder knobs to actually do something when the knobs turn (like Turn the device On/Off?), we could theoretically leverage "begin gesture" and "end gesture" events to show the (actively turning) encoder value on the bottom LCD row while the "gesture" hasn't ended yet, and then go back to displaying the second part of the device name after the "gesture" ends.

**Reply by markusschloesser** [reply]
> Although my MCU does not have any capacitive hardware

?? The faders are touch capacitive! 
The functionality I am talking about is

```
                if self._ChannelStripController__any_slider_is_touched():
                    display_parameters.append(slider_param)
```
Which is responsible for showing the fader volume



> The way I understand that idea is when a user begins turning an encoder knob, for example, a "gesture" begins and the gesture doesn't end until the user stops turning the knob

That would come in handy for the strange way Live does scrub_clip, but I have yet to try it out with a real project anyway



> For example, in Channel Strip mode, the "device row" encoder knobs don't do anything right? (just the encoder buttons take to you Device mode) and the "device row" display takes up both LCD rows for the mapped "device name". If we had a reason to map the "device row" encoder knobs to actually do something when the knobs turn (like Turn the device On/Off?),

That's a great idea!! But we can already do this with the current script by forwarding cc. Only problem might be to "find" the Device on/off parameter
